compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3359,7 +3370,33 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                 }
 
                 match arg.layout.abi {
-                    Abi::Aggregate { .. } => {}
+                    Abi::Aggregate { .. } => {
+                        // Pass and return structures up to 2 pointers in size by value,
+                        // matching `ScalarPair`. LLVM will usually pass these in 2 registers
+                        // which is more efficient than by-ref.
+                        let ptr_size = Pointer.size(self);
+                        let max_by_val_size = ptr_size * 2;
+                        let size = arg.layout.size;
+
+                        if arg.layout.is_unsized() || size > max_by_val_size {
+                            arg.make_indirect();
+                        } else if size > ptr_size && unlikely!(self.has_all_float(&arg.layout)) {
```

Comment:
Curious: Why the `unlikely!` hint?

Reply:
`unlikely!` hints can result in the function getting outlined even when inlining it would be profitable, thereby adding an indirection and making the call even slower (the first call, for instance, will always look a certain way).